internal/fs/kernel_list_cache_test.go

```diff
@@ -0,0 +1,353 @@
+}
+
+func (t *KernelListCacheTestCommon) SetupTest() {
+	t.createFilesAndDirStructureInBucket()
+	cacheClock.SetTime(time.Date(2015, 4, 5, 2, 15, 0, 0, time.Local))
+}
+
+func (t *KernelListCacheTestCommon) TearDownTest() {
+	t.fsTest.TearDown()
+}
+
+func (t *KernelListCacheTestCommon) TearDownSuite() {
+	t.fsTest.TearDownTestSuite()
+}
+
+func TestKernelListCacheTestSuite(t *testing.T) {
+	suite.Run(t, new(KernelListCacheTestCommon))
+}
+
+// createFilesAndDirStructureInBucket creates the following files and directory
+// structure.
+// bucket
+//
+//	explicitDir/
+//	explicitDir/file1.txt
+//	explicitDir/file2.txt
+//	implicitDir/file1.txt
+//	implicitDir/file2.txt
+func (t *KernelListCacheTestCommon) createFilesAndDirStructureInBucket() {
+	assert.Nil(t.T(), t.createObjects(map[string]string{
+		"explicitDir/":          "",
+		"explicitDir/file1.txt": "12345",
+		"explicitDir/file2.txt": "6789101112",
+		"implicitDir/file1.txt": "-1234556789",
+		"implicitDir/file2.txt": "kdfkdj9",
+	}))
+}
+
+func (t *KernelListCacheTestCommon) deleteObjectOrFail(objectName string) {
+	assert.Nil(t.T(), t.deleteObject(objectName))
+}
+
+type KernelListCacheTestWithPositiveTtl struct {
+	suite.Suite
+	fsTest
+	KernelListCacheTestCommon
+}
+
+func (t *KernelListCacheTestWithPositiveTtl) SetupSuite() {
+	t.serverCfg.ImplicitDirectories = true
+	t.serverCfg.MountConfig = &config.MountConfig{
+		ListConfig: config.ListConfig{
+			KernelListCacheTtlSeconds: kernelListCacheTtlSeconds,
+		}}
+	t.serverCfg.RenameDirLimit = 10
+	t.fsTest.SetUpTestSuite()
+}
+
+func TestKernelListCacheTestWithPositiveTtlSuite(t *testing.T) {
+	suite.Run(t, new(KernelListCacheTestWithPositiveTtl))
+}
+
+// (a) First ReadDir() will be served from GCSFuse filesystem.
+// (b) Second ReadDir() within ttl will be served from kernel page cache.
+func (t *KernelListCacheTestWithPositiveTtl) TestKernelListCache_CacheHit() {
+	// First read, kernel will cache the dir response.
+	f, err := os.Open(path.Join(mntDir, "explicitDir"))
+	assert.Nil(t.T(), err)
+	defer func() {
+		assert.Nil(t.T(), f.Close())
+	}()
+	names1, err := f.Readdirnames(-1)
+	assert.Nil(t.T(), err)
+	assert.Equal(t.T(), 2, len(names1))
+	assert.Equal(t.T(), names1[0], "file1.txt")
+	assert.Equal(t.T(), names1[1], "file2.txt")
+	err = f.Close()
+	assert.Nil(t.T(), err)
+	// Adding one object to make sure to change the ReadDir() response.
+	assert.Nil(t.T(), t.createObjects(map[string]string{
+		"explicitDir/file3.txt": "123456",
+	}))
+	defer t.deleteObjectOrFail("explicitDir/file3.txt")
+	// Advancing the clock within time.
+	cacheClock.AdvanceTime(kernelListCacheTtlSeconds * time.Second / 2)
+
+	// 2nd read, ReadDir() will be served from page-cache, that means no change in
+	// response.
+	f, err = os.Open(path.Join(mntDir, "explicitDir"))
+	assert.Nil(t.T(), err)
+	names2, err := f.Readdirnames(-1)
+
+	assert.Nil(t.T(), err)
+	assert.Equal(t.T(), 2, len(names2))
+	assert.Equal(t.T(), names1[0], names2[0])
```

Comment:
Nit: can we explicitly compare with constants/strings here instead of variables to make the test more readable. same for other tests